doc/guide/transactions-and-threading.rst

```diff
@@ -0,0 +1,394 @@
+thread-local transaction manager.
+
+Because we used a thread-local transaction manager, all of the work in
+the transaction needs to happen in the same thread.  Similarly, only
+one transaction can be active in a thread.
+
+If we want to have transactions who's work is spread over multiple
+threads, or if we wanted to run multiple simultaneous transactions in
+a single thread, then we can create transaction managers ourselves and
+pass them to :meth:`~ZODB.DB.open`::
+
+  tm = transaction.TransactionManager()
+  conn = db.open(tm)
+  conn.root.x = 2
+  tm.commit()
+
+.. -> src
+
+   >>> exec(src)
+
+In this example, to commit our work, we called ``commit()`` on the
+transaction manager we created and passed to :meth:`~ZODB.DB.open`.
+
+context managers
+----------------
+
+In the examples above, the transaction beginnings were
+implicit. Transaction's were effectively
+[#implicit-transaction-creation]_ created when the transaction
+managers were created and when previous transactions were committed.
+We can create transactions explicitly using
+:meth:`~transaction.interfaces.ITransactionManager.begin`::
+
+  tm.begin()
+
+.. -> src
+
+   >>> exec(src)
+
+A more modern [#context-managers-are-new]_ way to manage transaction
+boundaries is to use context managers and the Python ``with``
+statement. Transaction managers are context managers, so we can use
+them with the ``with`` statement directly::
+
+  with tm as trans:
+     trans.note("incrementing x")
+     conn.root.x += 1
+
+.. -> src
+
+   >>> exec(src)
+   >>> conn.root.x
+   3
+
+
+When used as a context manager, a transaction manager explicitly
+begins a new transaction, executes the code block and commits the
+transaction if there isn't an error and aborts it of there is an
+error.
+
+We used ``as trans`` above to get the transaction.
+
+Databases provide the :meth:`~ZODB.DB.transaction` to execute a code
+block in a transaction::
+
+  with db.transaction() as conn2:
+     conn2.root.x += 1
+
+.. -> src
+
+   >>> exec(src)
+
+Here, when we used ``as``, we got a connection, not a transaction.
+This is because a new connection is opened by the
+:meth:`~ZODB.DB.transaction`` method.  A new transaction manager was
+used as well.
+
+Getting a connection's transaction manager
+------------------------------------------
+
+In the previous example, you may have wondered how one might get the
+current transaction. Every connection has an associated transaction
+manager, which is available as the ``transaction_manager`` attribute.
+So, for example, if we wanted to set a transaction note::
+
+
+  with db.transaction() as conn2:
+     conn2.transaction_manager.get().note("incrementing x again")
+     conn2.root.x += 1
+
+.. -> src
+
+   >>> exec(src)
+   >>> db.history(conn.root()._p_oid)[0]['description']
+   'incrementing x again'
+
+Here, we used the
+:meth:`~transaction.interfaces.ITransactionManager.get` method to get
+the current transaction.
```

Comment:
WRT clearing the connection transaction manager, +1, less from a resource-management perspective than from a way to maybe avoid opportunities to fail, especially when a TM was a (local) global.

(It's funny to refer to a global as thread/greenlet local, but it makes sense because it's a thread/greenlet local global. :) )